Automattic/wp-calypso
# Purchase Management: Replace getUserPurchases with React Hook useUserPurchases

## Proposed Changes

This PR creates a new React Hook using useQuery called `useUserPurchases()` which fetches all of a user's purchases and returns them. This takes the place of the `getUserPurchases()` Redux function and its associated state.

This PR then uses the new Hook in the purchase management pages.

> [!NOTE]
> This does not remove the `QueryUserPurchases` component in the purchase list because other sub-components of the page might still rely on the Redux store being populated. 

> [!NOTE]
> There are a few other places that use `getUserPurchases()` within other parts of calypso but to keep this PR easy to review and test, I've limited it to only the main purchase list page. Follow-up PRs can migrate the other consumers and then eventually remove the Redux state itself.

## To do

The main issue with this approach currently is that the Redux store is a single place where three different mechanisms can populate the local cache with purchase data: fetching user purchases, fetching site purchases, and fetching a single purchase. By changing these components to use a hook, we no longer share that cache and instead will potentially have three separate caches that may all contain the same data.

- [ ] Replace some uses of `hasLoadedUserPurchasesFromServer` in `client/me/purchases` that may be affected by this change if they also depended on the query component.
- [ ] Make sure nothing in `client/me/purchases` uses the main `isDataLoading()` function.
- [ ] Replace some (all?) uses of `getByPurchaseId` in `client/me/purchases`.

## Why are these changes being made?

Using Redux is overkill for fetching a list of purchases and it limits these components to be used only within the main calypso package. useQuery (aka: React Query/Tanstack Query) is the main replacement for Redux currently used by calypso and this migration allows removing a lot of state from our Redux tree.

There is already a hook like this for site purchases (`useSitePurchases` added in https://github.com/Automattic/wp-calypso/pull/84761). This PR adds `useUserPurchases` alongside it in the `@automattic/data-stores` package.

## Testing Instructions

- Visit `/me/purchases` for an account with multiple purchases and verify that they show up and operate the same before and after this PR.
